source/containerimage/pull.go

```diff
@@ -173,6 +173,10 @@ func mainManifestKey(ctx context.Context, desc specs.Descriptor, platform specs.
 func (p *puller) CacheKey(ctx context.Context, g session.Group, index int) (cacheKey string, cacheOpts solver.CacheOpts, cacheDone bool, err error) {
 	p.Puller.Resolver = resolver.DefaultPool.GetResolver(p.RegistryHosts, p.Ref, "pull", p.SessionManager, g).WithImageStore(p.ImageStore, p.id.ResolveMode)
 
+	// progressFactory needs the outer context, the context in `p.g.Do` will
+	// be canceled before the progress output is complete
+	progressFactory := progress.FromContext(ctx)
```

Comment:
looking at `FromContext` implementation it only uses context for the `ctx.Value(contextKey)` call and later uses another context coming with the call. So why does it matter if this context gets canceled. Also, it is possible that this context also gets canceled before the blobs are actually pulled.

Reply:
yeah, it was not obvious to me either.

The progress factory is not used until later when we call the controller `Start`
https://github.com/moby/buildkit/blob/699121c00aeb45c066a0396a02669d29c4a4b9a7/util/progress/controller/controller.go#L30

Start is eventually called here when we need to Unlazy the ref:
https://github.com/moby/buildkit/blob/699121c00aeb45c066a0396a02669d29c4a4b9a7/cache/remote.go#L201
By the time we call Start, the original context from the `p.g.Do` has already been canceled, the new writer immediately gets ignored via:
https://github.com/moby/buildkit/blob/699121c00aeb45c066a0396a02669d29c4a4b9a7/util/progress/progress.go#L171-L172

Reply:
As discussed in slack, it seems the reader is what is getting associated with the canceled context via the flightcontrol on g.Do:
https://github.com/moby/buildkit/blob/699121c00aeb45c066a0396a02669d29c4a4b9a7/util/flightcontrol/flightcontrol.go#L105-L110

So this fix is really about using the reader associated with the CacheKey ctx